Add `CategoryBasis` for One-Hot Encoding of Discrete Variables (#527)
flatironinstitute/nemos: docs/how_to_guide/categorical_identifiability.md

Reviewer: I wonder if this notebook should be merged with the categorical_predictors one. It's basically answering the question: "why should I drop a column?", so it strikes me as unlikely that someone would come directly to this notebook. And, as written, it's a bit hard to read by itself. Or, at the very least, it shouldn't be at the same importance level as the categorical predictions one. You could do nesting (like how I do for the [texture model](https://docs.plenoptic.org/docs/branch/main/user_guide/models_and_metrics/portilla_simoncelli/ps_index.html)), to make it clear that these two are related and while the first one should be read by anyone who wants to use a categorical feature, the second one is really just to answer one specific question.

---
jupytext:
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.16.4
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

(categorical_identifiability)=
# Resolving Redundancy in Categorical Design Matrices

## Why Does Redundancy Arise?

When the [`Category`](nemos.basis.Category) basis is used as a standalone main-effect predictor

> together with a NeMoS GLM (which always includes an intercept), there are infinitely many coefficient vectors that produce the same linear predictor $X \cdot \mathbf{w} + c$. Let's see why:

Reviewer: this jumps right into math notation, which we don't often use in our docs. maybe some introduction before this section that says something like, "in this notebook, we're discussing a basic issue with categorical predictors in linear (and related) models. Thus, we'll discuss this problem using the following notation..." (and also link back to [this page](https://nemos--527.org.readthedocs.build/en/527/background/plot_00_conceptual_intro.html#glm-components), which reminds them of the relationship between this notation and the GLM)

```{code-cell} ipython3
import numpy as np
import nemos as nmo

category = np.array(["L", "L", "L", "L", "R", "R", "R", "R"])
cat_basis = nmo.basis.Category(["L", "R"])
X = cat_basis.compute_features(category)

print("One-hot encoding:\n", X)
print("Sum over columns:", X.sum(axis=1))
```

> Because the one-hot columns sum to a vector of ones, we can find coefficients $\mathbf{w}$ and
> intercept $c$ for which $X \cdot \mathbf{w} + c = 0$:

Reviewer: ```suggestion
Because the columns of our design matrix sum to a vector of ones, we can find coefficients $\mathbf{w}$ and intercept $c$ for which $X \cdot \mathbf{w} + c = 0$:
```

```{code-cell} ipython3
c = -1
w = np.array([1, 1])

print("X @ w + c:", X @ w + c)
```

> Stacking $c$ and $\mathbf{w}$ into a single vector $\mathbf{v} = [c, \mathbf{w}]$ and prepending

Reviewer: ```suggestion
This is bad! It means that our design matrix has a [null space](https://en.wikipedia.org/wiki/Kernel_(linear_algebra)), vectors we can add to the parameters that have no effect on the model predictions. This is what gives rise to the problem: our model cannot distinguish between different sets of parameters if their only differences lie in the null space. This both makes optimization more difficult and prevents us from interpreting the parameters.

Let's unpack this a bit more.

Stacking $c$ and $\mathbf{w}$ into a single vector $\mathbf{v} = [c, \mathbf{w}]$ and prepending
```

or something. I just think we need a bit more explanation before jumping into it

Reviewer: the information is all here in this section, I just think it needs to be fore-fronted a bi tmore

Author: I applied the suggestion, I like the emphasis on the practical issue too

> an all-ones column to $X$ gives the compact form $X_\text{aug} \cdot \mathbf{v} = 0$:

Reviewer: should say something like, "this representation is equivalent to the above, but more convenient for our purposes in this notebook"

```{code-cell} ipython3
v = np.hstack([c, w])
X_aug = np.column_stack([np.ones(len(category)), X])  # [1 | X]

print("X_aug @ v:", X_aug @ v)  # = 0  →  v is in the null space of X_aug
```

In mathematical terms, we can say that $\mathbf{v}$ is in the null space of the augmented matrix $X_\text{aug}$. This happens when at least one column of $X_\text{aug}$ can be expressed as a weighted sum of the other columns—that is, the columns are **linearly dependent**. The representation is therefore redundant, and we can add any multiple of $\mathbf{v}$ to the parameters without changing the predictions:

```{code-cell} ipython3
params = np.array([2., 0.5, -0.3]) # arbitrary parameters [intercept, coef]
alpha = 5.0                        # arbitrary shift along the null direction

print("X_aug @ params:             ", X_aug @ params)
print("X_aug @ (params + alpha*v): ", X_aug @ (params + alpha * v))
```

Models with parameters $[c, \mathbf{w}]$ and $[c, \mathbf{w}] + \alpha \cdot \mathbf{v}$ predict the same firing rate for any $\alpha \in \mathbb{R}$.
This is **non-identifiability**: the data alone cannot distinguish between them, and there is no unique optimal solution.

> A practical way to detect this redundancy is to check whether the rank of `X_aug = [1 | X]` is strictly less than number of its columns.

Reviewer: haven't explained the `[1|X]` notation yet, should mention it when we define `X_aug`

Reviewer: also, shouldn't that be latex, not monospace?

> **The rank is the number of linearly independent columns**, so if it is smaller than the total number of columns, some columns must be redundant. In that case, a non-trivial null space exists, i.e., there are non-zero vectors $\mathbf{v}$ such that $X_\text{aug} \cdot \mathbf{v} = 0$:

Reviewer: link to something about rank / linear independency here?

```{code-cell} ipython3
print(f"Rank of [1 | X]:   {np.linalg.matrix_rank(X_aug)}")
print(f"Number of columns: {X_aug.shape[1]}")

> # rank < n_cols  →  null space exists  →  model is non-identifiable

Reviewer: I think this comment belongs in the text

```

## Reference Coding: A Simple Way to Recover Identifiability

> Drop one column from the full encoding. The retained columns become *contrasts* against the

Reviewer: ```suggestion
One way to solve this problem, known as "reference coding", is to drop one column from the full encoding. The retained columns become *contrasts* against the
```

dropped (reference) category. All columns are now linearly independent:

```{code-cell} ipython3
X_ref = X[:, 1:]  # drop "L"; the remaining column codes "R" vs "L"
X_ref_aug = np.column_stack([np.ones(len(category)), X_ref])
# rank == n_cols → full rank
print(f"Rank of [1 | X_ref]: {np.linalg.matrix_rank(X_ref_aug)}")
print(f"Number of columns:   {1 + X_ref.shape[1]}")
```

> The reference level is arbitrary from a model-fit perspective, but determines coefficient

Reviewer: ```suggestion
Which column you drop is arbitrary from a model-fit perspective, but determines coefficient
```
